### ISSUE TITLE
git autoversioning can be confused if mosh tarball extracted in another git directory

### ISSUE THREAD

**Issue by achernya** [keithw/mosh#646]
While working on building 1.2.4.95rc2 for Fedora, I ran into a problem with `make distcheck` after extracting the mosh tarball inside of my Fedora packaging directory (which itself is managed in git). When trying to compute the git version, the rule fails.

I believe we can make the git autoversioning more robust if we make it check for the presence of .git in the top-level directory before doing these operations.

**Comment by keithw** [comment]
Just to make sure I understand -- where did this tarball come from? The GitHub auto-tarballification from the tag, or from a previous "make dist", or...?

I think we basically do not want to advertise the GitHub auto-tarballized tgz ever. (We will post a GNU-style mosh-1.2.5.tar.gz on mosh.mit.edu from "make distcheck".)

But I agree that there's still a chance for confusion here so this fix sounds like a good idea.

**Comment by achernya** [comment]
This was while I was trying to run `make distcheck` from the github auto-tarballification to create a proper dist tarball so I could feed it to rpmbuild. I had extracted it in my fedora/mosh directory.

**Comment by cgull** [comment]
Yes, that's an issue.  Git keeps surprising us :) so I'm going to think about this a little bit and see if I can come up with something more general than `test -d .git`...

**Comment by keithw** [comment]
How do you feel about just adding a fallthrough to "unknown" if the version check fails for any reason?

**Comment by cgull** [comment]
That, and fixing the code so that it uses the same Git command to test
for Git *and* produce a version, and adding the "--always" flag (for
shallow repos) are all reasonable things.  Let's figure out what makes
sense here.

IRC, maybe?

**Comment by cgull** [comment]
So I think this resolves these issues (and 3 more-- I must have had a bad day when I wrote this).  Tested on FreeBSD and OS X, I'll be checking other platforms tonight.  @keithw, I chose to fall back to the distribution-generated version, rather than "unknown".

I'd be OK with trimming this back to just the git-in-condition-expressions part of it, or something else simple, given that we're trying to get this out the door.
